### PR TITLE
Scan iterators (v4)

### PR DESCRIPTION
**Pull request by dim**
Here you go, as promised in #303

### CODE REVIEW

File: example_test.go [hunk 1]
```diff
@@ -314,3 +314,58 @@ func Example_customCommand() {
 	fmt.Printf("%q %s", v, err)
 	// Output: "" redis: nil
 }
+
+func ExampleScanIterator() {
+	iter := client.ScanIterator(nil)
+
+	for iter.Next() {
+		fmt.Println(iter.Value())
```
Comment (vmihailenco) [inline example_test.go]: I would rename `Value` to `Key`.

**Comment by vmihailenco** [comment]
Did you consider following API:

```
scan := client.Scan(0, "", 100).Result()
for scan.Next() {
    fmt.Println(scan.Key())
}
fmt.Println(scan.Err())
```

or


```
scan := client.Scan(0, "", 100).Iterator()
for scan.Next() {
    fmt.Println(scan.Key())
}
fmt.Println(scan.Err())
```

File: example_test.go (hunk 1 above)
Comment (dim) [inline example_test.go]: The trouble is that it's `Key` for `SCAN` and `Field` for all the others. I rename it to `Val` to keep it consistent with `Cmd`

**Comment by dim** [comment]
So yes, I did, but the trouble is that you have to parse a cursor value, which isn't exactly intuitive. Happy to adjust, just not sure it's the best option. Let me know ...

**Comment by dim** [comment]
@vmihailenco PTAL ^^

File: command.go [hunk 2]
```diff
@@ -685,11 +685,22 @@ func (cmd *ZSliceCmd) readReply(cn *pool.Conn) error {
 
 //------------------------------------------------------------------------------
 
+// Scanner is a ScanCmd with Iterator capabilities
+type Scanner struct {
+	*ScanCmd
+	c *commandable
+}
```
Comment (vmihailenco) [inline command.go]: Why not to remove `Scanner` and add `Iterator() *ScanIterator` method to `ScanCmd`?

File: iterator.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+	// Return if there is more data to fetch
+	if i.scn.cursor == 0 {
+		return false
+	}
+
+	// Fetch next page
+	i.scn._args[1] = i.scn.cursor
+	i.scn.reset()
+	i.scn.c.Process(i.scn.ScanCmd)
```
Comment (vmihailenco) [inline iterator.go]: Nice trick!

Comment (vmihailenco) [inline iterator.go]: There is also opportunity to improve var names, e.g.:
- it.cmd._args
- it.client.Process

File: command.go (hunk 2 above)
Comment (dim) [inline command.go]: So that would require `ScanCmd` to be able to have `commandable` too as it would need to pass it to `ScanIterator`

Comment (vmihailenco) [inline command.go]: Works for me. It is essentially the same, but now you also have additional type Scanner for this...

Comment (dim) [inline command.go]: Yes, but there is a public `NewScanCmd` API which we would need to deprecate.

Comment (vmihailenco) [inline command.go]: We can do it in v4, but I don't see why you need it. Just add new private field to existing struct...

Comment (vmihailenco) [inline command.go]: We can also change API to `client.Scan(0, "", 20).Iterator(client)`

**Comment by vmihailenco** [comment]
@dim I added some changes. PTAL

File: command.go (hunk 2 above)
Comment (dim) [inline command.go]: That seems clumsy to me, that' why I added a new type wrapping ScanCmd. When client is the origin of your command, then you have an `Iterator` method, but you can also construct your own `ScanCmd` with `NewScanCmd` if you really want to. For me, the additional type is worth the flexibility. I am happy to rename `Scanner` to `ScanCmdIterable` or something more verbose.

**Comment by vmihailenco** [comment]
@dim restored `Scanner`, PTAL

**Comment by dim** [comment]
LGTM

**Pull request merged**